compiler/transforms/convert_linalg_to_acc.py

```diff
@@ -0,0 +1,237 @@
+                        state[accel] = op.out_state
+                    else:
+                        pass
+                elif isinstance(op, scf.If):
+                    # grab the computed state for both sides:
+                    if_state = _walk(op.true_region, state.copy(), rewriter)
+                    else_state = _walk(op.false_region, state.copy(), rewriter)
+
+                    # calculate the delta:
+                    delta = calc_if_state_delta(state, if_state, else_state)
+                    # no delta = nothing to do
+                    if not delta:
+                        continue
+
+                    # grab a list of added return vals for both branches
+                    new_vals: tuple[tuple[SSAValue, ...], ...] = (
+                        tuple(if_val for if_val, _ in delta.values()),
+                        tuple(else_val for _, else_val in delta.values()),
+                    )
+                    # for each branch, rewrite the yield to return the new state
+                    for branch, added_vals in zip(op.regions, new_vals):
+                        assert isinstance(branch, Region)
+                        if not branch.blocks:
+                            branch.add_block(Block([scf.Yield(*added_vals)]))
+                        else:
+                            assert (
+                                branch.block.last_op is not None
+                            )  # we know there is a yield op
+                            rewriter.replace_op(
+                                branch.block.last_op,
+                                scf.Yield(*branch.block.last_op.operands, *added_vals),
+                            )
+                    # then, insert a new if with additional return values:
+                    num_scf_results = len(op.results)
+                    rewriter.replace_op(
+                        op,
+                        new_if := scf.If(
+                            op.cond,
+                            [val.type for val in (*op.results, *new_vals[0])],
+                            op.detach_region(op.regions[0]),
+                            # is fine because previous line removes a region
+                            op.detach_region(op.regions[0]),
+                        ),
+                        [new_if.results[i] for i in range(num_scf_results)],
+                    )
+                    # update state:
+                    for res in new_if.results[num_scf_results:]:
+                        assert isinstance(res.type, acc.StateType)
+                        state[res.type.accelerator.data] = res
+
+                # calling another function invalidates all states
+                elif isinstance(op, func.Call):
+                    state.clear()
+                elif isinstance(op, func.Return | scf.Yield):
+                    continue
+                else:
+                    raise RuntimeError(f"What is a {op}?")
+    return state
```

Comment:
Important change

Reply:
This pattern connects all the generated `acc2` calls via def-use chain such that we can optimize them later on